### PR TITLE
Improved conversion between `pyarrow` and `pandas` in P2P shuffling

### PR DESCRIPTION
**Pull request by hendrikmakait**
Addresses #7880
~Incorporates and blocked by #7895~

Performance seems to vary wildly between runs on the reproducer from #7880, but not converting to `string[python]` and back to `string[pyarrow]` generally seems like a good idea.

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by mrocklin** [comment]
Oh cool.  So we'll avoid ever making python object strings now?

**Comment by hendrikmakait** [comment]
> Oh cool. So we'll avoid ever making python object strings now?

We'll avoid creating them _per default_. If there's a column that uses `string[python]` in `meta`, it will still get converted in `return df.astype(meta.dtypes)`.

**Comment by mrocklin** [comment]
Sure.  That makes sense.  I'm curious to see how this impacts performance on current shuffling workloads.  Both due to just general performance, but also possibly complex GIL-interactions with network bandwidth.

Do we currently use `string[pyarrow]` in shuffling benchmarks?  If so, I'd be curious to see how this impacts performance.

File: distributed/shuffle/_arrow.py
Comment (phofl) [inline distributed/shuffle/_arrow.py:64]: maybe including ``pa.large_string()`` here?

Comment (hendrikmakait) [reply distributed/shuffle/_arrow.py]: Good point, done.

**Comment by phofl** [comment]
I've never used ``self_destruct=True`` personally, not sure how this impacts the arrays backing the table. Just mentioning that ``string[pyarrow]`` is zero-copy and uses the arrays from the table.

**Review by jrbourbeau — COMMENTED** [review]
Thanks @hendrikmakait -- LGTM

There are a bunch of shuffle-related test failures here. Not fully convinced they're actually related to the changes here since they're array rechunking tests (my guess is https://github.com/dask/distributed/pull/7856), but am holding off on merging until you get a chance to look at things.

**Comment by hendrikmakait** [comment]
@jrbourbeau: The test failures disappeared after merging with the latest `main` and skipping caches. This should be ready to merge.

**Review by jrbourbeau — APPROVED** [review]
Thanks @hendrikmakait! This is in

**Pull request merged**
